test/self_protocore/upsert_validator.js

```diff
@@ -0,0 +1,151 @@
+// Copyright 2020 OpenST Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+const BN = require('bn.js');
+
+const { AccountProvider } = require('../test_lib/utils.js');
+const Utils = require('../test_lib/utils.js');
+
+const TestSelfProtocore = artifacts.require('TestSelfProtocore');
+
+const config = {};
+const validator = {};
+
+contract('SelfProtocore::upsertValidator', (accounts) => {
+  const accountProvider = new AccountProvider(accounts);
+
+  beforeEach(async () => {
+    config.coconsensusAddress = accountProvider.get();
+
+    config.epochLength = new BN(100);
+
+    config.genesisKernelHeight = new BN(1);
+    config.genesisKernelHash = Utils.getRandomHash();
+
+    config.core = accountProvider.get();
+    config.genesis = {};
+    config.genesis.auxiliaryParentVoteMessageHash = Utils.getRandomHash();
```

Comment:
this should either move into a local `utils.js`, or we can have `SelfProtocoreValidatorSetMock` that bypasses these params

Reply:
this can be improved later

Reply:
does not need to be done now; but add a @TODO, or @jayeshbairagi how should we track "known areas of improvements"

Reply:
Please take a reference from https://github.com/mosaicdao/mosaic-1/pull/244 so as to have minimum conflicts.

Reply:
Done, noted for further discussion